[MIG][8.0] module_auto_update backport from 9

sbidoul opened the pull request:
Last commit is migration, removing the deprecated features that never existed in 8.

hughesbm committed:
[ADD] module_auto_update: Create module (#882)

* [IMP] module_auto_update: Create new module
* Add checksum_dir and checksum_installed fields to ir.module.module
  * Add checksum_dir to compute current checksum of module directory
    in addons path
  * Add checksum_installed to store checksum of module directory
    when module was last installed or upgraded
* Use checksumdir Python library to compute module directory sha1
  hashes, ignoring pyc and pyo extensions
* Extend update_list method to compare modules' checksum_dir and
  checksum_installed, then change state of modules with differing
  checksums to 'to upgrade'
* Replace Apps/Updates menu item with menu item of same name, which
  updates apps list and displays tree view of ir.module.module
  records with state 'to upgrade'
* Extend create and write methods to store computed checksum_dir as
  checksum_installed during module installation and upgrade, and
  set checksum_installed to False on uninstall
* Use context to stop checksum_installed from being updated during
  upgrade/uninstall cancellation
* Add cron job to periodically check for module upgrades by
  comparing checksums, then perform any available upgrades
* Extend upgrade_module method (called by cron and 'Apply Scheduled
  Upgrades' menu item) to call update_list
* Add post_init_hook to store checksum_installed of existing
  modules
* Add test coverage

* [FIX] module_auto_update: Fix test broken by changes
* Use dummy module to test update_list method instead of
  module_auto_update

oca-transbot committed:
OCA Transbot updated translations from Transifex

yajo committed:
[MIG][module_auto_update] Backport from v10

yajo committed:
[REF][module_auto_update] Recompute dir hashes only when needed

By removing the recomputation from `update_list` we get faster CLI module upgrades and it only performs the autoupdate when using the autoupdate wizard or cron.

yajo committed:
[FIX][module_auto_update] Pass tests if addon is in readonly directory

yajo committed:
[FIX][module_auto_update] Set dependencies to upgrade

pedrobaeza committed:
[FIX] module_auto_update: Missing author

oca-transbot committed:
OCA Transbot updated translations from Transifex

yajo committed:
[FIX][module_auto_update] Record base addon checksum (#948)

yajo committed:
[FIX][module_auto_update] Always store changes in lower graphs (#953)

oca-transbot committed:
OCA Transbot updated translations from Transifex

sbidoul committed:
[REF] module_auto_update: Step 2, add new API

This code comes from the module_checksum_upgrade proposal
at https://github.com/OCA/server-tools/pull/1176.

* [ADD] module_checksum_upgrade

It provides the core mechanism of module_auto_update without
the cron nor any change to the standard upgrade mechanism.
Instead it provides an API on which module_auto_update can build,
as well as a method which can be called from a script to run
the upgrade of modules for which the checksum has changed.

* [IMP] refactor module_auto_update

Make it depend on module_checksum_upgrade which provides
the core mechanisms of managing the checksums. module_auto_update
makes it automatic.

* [IMP] module_checksum_upgrade: better exclusion mechanism

Ignore files based on exclude patterns.
Ignore uninstalled languages.
Better default for patterns to ignore (*.pyc,*.pyo,*.pot,static/*)

For better control on the hashing mechanism implement our own:
it's quite easy, and the checksumdir module used previously had
no test.

* [MIG] module_auto_update: adapt to new checksum mechanism

* [IMP] module_checksum_upgrade: raise in case of
 incomplete upgrade

* [IMP] module_checksum_upgrade: improve default exclusion
 pattern

* [IMP] module_checksum_upgrade: control translations
 overwrite

* [IMP] module_checksum_upgrade: one more test

* [IMP] module_checksum_upgrade: credits [ci skip]

yajo committed:
[REF] module_auto_update: Step 1, move all deprecated stuff to deprecated files

- Files are clearly suffixed with `_deprecated` so we know those features have no support nor migrations.
- Views are removed, since updating from UI was too buggy to support it anymore.

yajo committed:
[REF] module_auto_update: Step 3, backwards compatibility

The previous implementation of this addon proved being extremely buggy:

- It supplied out of the box a enabled cron to update Odoo that didn't restart the server, which possibly meant that upgrades broke things.
- It overloaded standard Odoo upgrade methods that made i.e. installing an addon sometimes forced to upgrade all other addons in the database.
- The checksum system wasn't smart enough, and some files that didn't need a module upgrade triggered the upgrade.
- It was based on a dirhash library that was untested.
- Some updates were not detected properly.
- Storing a column into `ir.module.module` sometimes forbids uninstalling the addon.

Thanks to Stéphane Bidoul (ACSONE), now we have new methods to perform the same work in a safer and more stable way.

All I'm doing here is:

- Cron is disabled by default.
- Installed checksums are no longer saved at first install.
- Old installations should keep most functionality intact thanks to the migration script.
- Drop some duplicated tests.
- Allow module uninstallation by pre-removing the fields from ir.mode.model.
- When uninstalling the addon, the deprecated features will get removed for next installs always.

Besides that, fixes for the new implementation too:

- When uninstalling the addon, we remove the stored checksum data, so further installations work as if the addon was installed from scratch.

sbidoul committed:
[MIG] module_auto_update from 9 to 8

JayVora-SerpentCS commented:
Better to merge OCA transbot commits into 1

pedrobaeza commented:
In this case they are not adjacent, so they can be let as is.

yajo commented on module_auto_update/models/module_deprecated.py:
```diff
@@ -1,69 +1,4 @@
 
-from openerp import api, fields, models
-
 PARAM_DEPRECATED = "module_auto_update.enable_deprecated"
```
I think you can remove this file and declare the param name directly into the hooks file; it's now used only there.

pedrobaeza replied:
We don't need deprecated features in v8 as there was no module there, isn't it? But if it's more work, it can be keep.

sbidoul replied:
@pedrobaeza the deprecated features are not here. The parameter is required here in 8, to avoid the deprecated features are enabled automatically when migrating to 9.

sbidoul replied:
@yajo is that blocking for you?

yajo replied:
Well... not blocking, I approved already. However, are there any reasons to keep this file just for setting a variable?

sbidoul replied:
Well it's just in the same place as in the other versions.

yajo commented:
Please include fix from #1205

yajo committed:
[FIX] module_auto_update: Add .pyo sample file (#1205)

Without this patch, if your tests are run under a `PYTHONOPTIMIZE=2` precompiled environment, they'd fail with this error because a new `.pyo` file would be created.:

    FAIL: test_basic (openerp.addons.module_auto_update.tests.test_addon_hash.TestAddonHash)
    Traceback (most recent call last):
    `   File "/opt/odoo/auto/addons/module_auto_update/tests/test_addon_hash.py", line 41, in test_basic
    `     'static/src/some.js',
    ` AssertionError: Lists differ: ['README.rst', 'data/f1.xml', ... != ['README.rst', 'data/f1.xml', ...
    `
    ` First differing element 13:
    ` models/stuff.pyo
    ` static/src/some.js
    `
    ` First list contains 1 additional elements.
    ` First extra element 14:
    ` static/src/some.js
    `
    `   ['README.rst',
    `    'data/f1.xml',
    `    'data/f2.xml',
    `    'i18n/en.po',
    `    'i18n/en_US.po',
    `    'i18n/fr.po',
    `    'i18n/fr_BE.po',
    `    'i18n/test.pot',
    `    'i18n_extra/en.po',
    `    'i18n_extra/fr.po',
    `    'i18n_extra/nl_NL.po',
    `    'models/stuff.py',
    `    'models/stuff.pyc',
    ` -  'models/stuff.pyo',
    `    'static/src/some.js']
    Ran 3 tests in 0.005s
    FAILED

With this patch, the `.pyo` file is included, so tests will pass anywhere.

sbidoul commented:
@Yajo done
